MdeModulePkg/Include/Library/DeviceStateLib.h

```diff
@@ -32,27 +32,71 @@ MU_CHANGE: new file
 
 #define DEVICE_STATE_MAX  (1 << (31))
 
+#define DEVICE_STATE_PLATFORM_MODE_INDEX_START  20
+#define DEVICE_STATE_PLATFORM_MODE_INDEX_END    27
+
+#define MAX_INSECURE_DEVICE_STATE_STRING_SIZE  512
+
+//
+// DEFINE the device state strings here. These are used for logging and measuring purposes.
+//
+#define DEVICE_STATE_SECUREBOOT_OFF_STRING             "SECURE_BOOT_OFF "
+#define DEVICE_STATE_MANUFACTURING_MODE_STRING         "MANUFACTURING_MODE "
+#define DEVICE_STATE_DEVELOPMENT_BUILD_ENABLED_STRING  "DEVELOPMENT_BUILD_ENABLED "
+#define DEVICE_STATE_SOURCE_DEBUG_ENABLED_STRING       "SOURCE_DEBUG_ENABLED "
+#define DEVICE_STATE_UNDEFINED_STRING                  "UNDEFINED "
+#define DEVICE_STATE_UNIT_TEST_MODE_STRING             "UNIT_TEST_MODE "
+#define DEVICE_STATE_PLATFORM_MODE_STRING              "OEM_DEFINED "
+
 typedef UINT32 DEVICE_STATE;
 
 /**
-Function to Get current device state
-@retval the current DEVICE_STATE
-**/
+ * Get String that represents the device's current insecure states, using the platform defined set of
+ * insecure device states. This can be used to measure insecure device states into the TPM or perform
+ * other required platform actions when the device enters an insecure state.
+ *
+ * @param[out] Buffer - Buffer to store the string
+ * @param[in] MaxSize - Maximum size of the buffer
+ * @retval The The number of characters in the string, including the null character.

```

Comment:
The function behavior when an invalid parameter is given should be defined.

For example:

- What is the expected behavior if `Buffer` is `NULL`?
- What is the expected behavior if `MaxSize` is `0`?

Reply:
Changed the function prototype to return EFI_STATUS. AddDeviceState uses RETURN_STATUS but is an EFIAPI - should that be changed as well?

Also, what do you think of the IN/OUT UINTN* for StringSize? Presumably the caller can just call AsciiStrnSizeS themselves?